R/epidatacall.R

```diff
@@ -183,6 +184,21 @@ fetch_classic <- function(epidata_call, fields = NULL, disable_data_frame_parsin
 
   response <- request_impl(epidata_call, "classic", fields)
   response_content <- httr::content(response, as = "text", encoding = "UTF-8")
+
+  # TODO Temporary workaround the first row of the response being a comment
+  # Remove on 2023-06-21
+  if (grepl("anonymous limit", response_content) && !epidata_call$only_supports_classic) {
+    rlang::warn("epidata error: anonymous user limit exceeded", "epidata_error")
+    response_content <- str_c("[", str_sub(response_content, 494, length(response_content) - 40), "]")
+    response_content <- jsonlite::fromJSON(response_content, simplifyDataFrame = !disable_data_frame_parsing)
+    return(response_content)
+  } else if (grepl("rate limit", response_content) && !epidata_call$only_supports_classic) {
+    rlang::warn("epidata error: rate limit exceeded", "epidata_error")
+    response_content <- str_c("[", str_sub(response_content, 489, length(response_content) - 40), "]")
+    response_content <- jsonlite::fromJSON(response_content, simplifyDataFrame = !disable_data_frame_parsing)
+    return(response_content)
+  }
```

Comment:
I think the idea of this phase is to force users to take some action.  An R warning doesn't do that, and is what we already had in Phase 1.  I suggest we just make this hard error but point to the escape hatch.  Something like the following.  (stringr imports could be removed if we use this.)

```suggestion
  if (grepl("This request exceeded", response_content) && !epidata_call$only_supports_classic) {
    response_content <- jsonlite::fromJSON(response_content, simplifyDataFrame = FALSE)
    message <- response_content$epidata[[1L]]
    cli::cli_abort(c(
      "epidata warning, promoted to error: {message}",
      "i" = "Either:",
      "*" = "set the environment variable DELPHI_EPIDATA_KEY, or",
      "*" = "set the option 'delphi.epidata.key':",
      " " = "{.code options(delphi.epidata.key = 'YOUR_KEY_OR_TEMPORARY_KEY')}",
      "To save your key for later sessions (and hide it from your code), you can edit your .Renviron file with:",
      "*" = "usethis::edit_r_environ()"
    ))
  }
```

This gives, e.g.,:
```
Error in `fetch_classic()` at epidatr/R/epidatacall.R:158:2:
! epidata warning, promoted to error: This request exceeded the rate limit on
  anonymous requests, which will be enforced starting 2023-06-21. To be exempt from
  this limit, authenticate your requests with a free API key, now available at
  https://api.delphi.cmu.edu/epidata/admin/registration_form. A temporary public key
  `TEMP-API-KEY-EXPIRES-2023-06-28` is available for use between now and 2023-06-28
  to give you time to register or adapt your requests without this message
  continuing to break your systems.
ℹ Either:
• set the environment variable DELPHI_EPIDATA_KEY, or
• set the option 'delphi.epidata.key':
  `options(delphi.epidata.key = 'YOUR_KEY_OR_TEMPORARY_KEY')`
To save your key for later sessions (and hide it from your code), you can edit your
.Renviron file with:
• usethis::edit_r_environ()
```

Reply:
Sounds good. Guess it's time to add an API key to this repo for CI.